### PR TITLE
Overlay: fix closing after click on item in shadow dom (T1146455)

### PR DESCRIPTION
**Pull request by Krijovnick**

### CODE REVIEW

**Review by marker-dao — DISMISSED** [review]

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 1]
```diff
@@ -855,6 +855,30 @@ testModule('visibility', moduleConfig, () => {
         }
     });
 
+    if(!QUnit.isInShadowDomMode()) {
+        test('overlay should not close after click on item in shadow dom (T1146455)', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/overlay.tests.js]: Let's specify that we're talking only about element inside an overlays's content. Overlay should still close after click on element in shadow dom if this element is not inside a content.
```suggestion
        test('overlay should not close after click on content element in shadow dom (T1146455)', function(assert) {
```

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 2]
```diff
@@ -855,6 +855,30 @@ testModule('visibility', moduleConfig, () => {
...
+            const $template = $('<div id=\'text\'></div>')
+
+            const overlay = $('#overlay').dxOverlay({
+                hideOnOutsideClick: true,
+                contentTemplate: new Template($template),
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/overlay.tests.js]: `new Template()` is a kind of a private api. We try to not use a private api to simplify a future refactoring.
Other points:
- `Template` api using makes the test more difficult. We all know that `contentTemplate` accepts a function returning a node, but not everybody knows an integration api
- `Template` api makes the test more synthetic. User cannot pass `Template` instance directly. It's more sensitive to test a real scenario.

I suggest u refactor the following way, we can also simplify getters.
```javascript
    const content = $('<div>').get(0);
    const overlay = $('#overlay').dxOverlay({
        // ...
        contentTemplate: () => content
    }).dxOverlay('instance');
    
    content.attachShadow({ mode: 'open' });
    content.shadowRoot.innerHTML = '<p>Inner Text</p>';
    
    const shadowTextElement = content.shadowRoot.querySelector('p');
```

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 3]
```diff
@@ -855,6 +855,30 @@ testModule('visibility', moduleConfig, () => {
         }
     });
 
+    if(!QUnit.isInShadowDomMode()) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/overlay.tests.js]: Why don't we want run this test in shadowDom? Seems it works fine

File: js/ui/overlay/ui.overlay.js [hunk 4]
```diff
@@ -303,15 +303,19 @@ const Overlay = Widget.inherit({
         this._initPositionController();
     },
 
+    _isTargetInContainer: function(container, target) {
+        return contains(container, target) || contains(container, target.getRootNode().host);
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: Hm, this code is not related to an overlay specific => it should be placed in some common core place.
Let's try to update initial `contains` method code?

**Review by marker-dao — DISMISSED** [review]

File: js/core/utils/dom.js [hunk 5]
```diff
@@ -93,7 +93,11 @@ export const contains = (container, element) => {
     if(!element) {
         return false;
     }
+    const shadowHost = element.getRootNode && element.getRootNode().host;
```
Comment (ksercs) [inline js/core/utils/dom.js]: If we change this code, we should write a new test in `utils.dom.tests.js`

File: js/core/utils/dom.js [hunk 6]
```diff
@@ -93,7 +93,11 @@ export const contains = (container, element) => {
...
+    return isInContainer(container, element) || shadowHost && isInContainer(container, shadowHost);
+};
 
+const isInContainer = (container, element) => {
```
Comment (ksercs) [inline js/core/utils/dom.js]: Let's move this private function to the file's beginning

File: js/core/utils/dom.js [hunk 7]
```diff
@@ -93,7 +93,11 @@ export const contains = (container, element) => {
...
     if(domAdapter.isTextNode(element)) {
```
Comment (ksercs) [inline js/core/utils/dom.js]: If we think that all this conditions are necessary for the case when target is a shadow host, it's a good idea to add tests for this cases

File: js/core/utils/dom.js [hunk 8]
```diff
@@ -93,22 +93,10 @@ export const contains = (container, element) => {
...
-        return container.documentElement.contains(element);
-    }
-
     if(isWindow(container)) {
         return contains(container.document, element);
     }
-
-    return container.contains
-        ? container.contains(element)
-        : !!(element.compareDocumentPosition(container) & element.DOCUMENT_POSITION_CONTAINS);
```
Comment (ksercs) [inline js/core/utils/dom.js:111]: Why did u remove this code?

File: testing/tests/DevExpress.core/utils.dom.tests.js [hunk 9]
```diff
@@ -118,3 +118,39 @@ QUnit.test('it correctly works with svg elements', function(assert) {
...
+    const $div = $('<div>').appendTo('#qunit-fixture');
+    const divContent = $div.get(0);
+
+    divContent.attachShadow({ mode: 'open' });
+    divContent.shadowRoot.innerHTML = '<p>Inner Text</p>';
+
+    const textElement = divContent.shadowRoot.querySelector('p');
+
+    assert.ok(domUtils.contains(document, textElement));
+});
```
Comment (ksercs) [inline testing/tests/DevExpress.core/utils.dom.tests.js]: empty line

File: testing/tests/DevExpress.core/utils.dom.tests.js [hunk 10]
```diff
@@ -118,3 +118,39 @@ QUnit.test('it correctly works with svg elements', function(assert) {
 
     assert.ok(domUtils.contains(svgContainer, childElement));
 });
+
+QUnit.test('element in shadow dom should detects in window element', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.core/utils.dom.tests.js]:
```suggestion
QUnit.test('element in shadow dom should be detected if container is window', function(assert) {
```

File: js/core/utils/dom.js (hunk 8 above)
Comment (Krijovnick) [reply js/core/utils/dom.js:111]: Because it was made for IE (https://github.com/DevExpress/DevExtreme/pull/13136). Now we does not support IE

Comment (ksercs) [reply js/core/utils/dom.js:111]: Nice!

**Review by ksercs — APPROVED** [review]

**Review by alexanderPolosatov — APPROVED** [review]

**Review by marker-dao — APPROVED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]
